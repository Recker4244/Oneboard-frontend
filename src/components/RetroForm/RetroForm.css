.retro-form {
  background-color: rgba(0, 0, 0, 0.734);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  position: fixed;
}

.retro-form-body-right h1 {
  font-size: 25px;
  font-weight: 10px;
  padding-left: 30px;
}

.retro-form-content {
  background-color: white;
  width: 60%;
  height: 75%;
}

.upload-images {
  display: flex;
  flex-direction: row;
}

.upload-images h2 {
  margin-left: 20px;
  width: 92%;
  font-size: 25px;
}

.label-container label {
  font-size: 20px;
  font-weight: 500;
  color: grey;
  padding-left: 30px;
  margin: 0px;
}
.upload-image {
  display: flex;
  gap: 10px;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 378px;
  width: 94%;
  padding: 20px;
  border-radius: 5px;
  border: 2px dashed #555;
  color: #444;
  cursor: pointer;
  transition: 0.2s ease-in-out, border 0.2s ease-in-out;
  margin: 0px;
  margin-top: 10px;
}

.upload-image:hover {
  background: #eee;
  border-color: #111;
}

.upload-image p {
  font-size: large;
}

.retro-form-body {
  display: flex;
  flex-direction: row;
  padding: 30px;
}


.retro-form-title {
  display: flex;
  flex-direction: row;
  background-color: #f7fbff;
  padding: 10px;
  justify-content: space-between;
}

.retro-form-title h1 {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.title-input {
  border-bottom: 1.5px;
  border-color: #030c82;
  border-style: solid;
  border-radius: 1px;
  margin-left: 29px;
  margin-right: 10px;
  width: 30%;
  height: 35px;
  padding-left: 10px;
  outline: none;
}

.action-item-input {
  border-bottom: 1.5px;
  border-color: #030c82;
  border-style: solid;
  border-radius: 1px;
  margin-left: 20px;
  margin-right: 10px;
  width: 59%;
  height: 35px;
  padding-left: 10px;
  outline: none;
}

.action-list {
  list-style: circle;
  list-style-type: disc;
  cursor: pointer;
  font-size: larger;
}
.action-lists :hover {
  color: #f24727;
}
.action-lists :hover{
    color: red;
}
.action-lists :hover{
    color: red;
}

.action-lists {
  padding-left: 20px;
  margin-left: 36px;
  margin-top: 10px;
}
.retro-form-body-right {
  margin-top: 10px;
  margin-left: 35px;
}

.add-title-button {
  margin-right: 40px;
}

.retro-form-title-left {
  display: flex;
  flex-direction: row;
  align-items: center;
}
.cancel-button {
  position: relative;
  right: 19px;
  font-size: 25px;
}
.label-container {
  border-right: gray;
  border-right-style: solid;
  border-right-width: 1px;
  padding: 10px;
  width: 60%;
}
.submit-btn {
  margin-top: 10px;
  cursor: pointer;
  font-weight: 500;
  padding: 11px 28px;
  margin-left: 45px;
  border-radius: 5px 5px;
  font-size: 0.8rem;
  border: none;
  color: #fff;
  background: #051c2c;
  margin-right: 10px;
  width: 250px;
}

.re-upload {
  display: flex;
  margin-top: 10px;
}
.retro-form-body img {
  width: auto;
  height: 90%;
  display: flex;
}
.re-upload-image {
  width: 100%;
  display: flex;
  height: 100%;
  object-fit: cover;
  flex-direction: column;
  align-items: center;
}
.retro-form-body-right button {
  font-weight: 500;
  color: #030c82;
  padding-top: 20px;
  padding-left: 40px;
}
.retro-form-body-right input {
  font-weight: 500;
  color: rgb(43, 43, 111);
  margin-top: 20px;
  margin-left: 35px;
  border-radius: 1px;
}
.submit-btn:hover {
  background-color: rgb(68, 176, 176);
}

.upload-image input {
  display: none;
}

label {
  cursor: pointer;
}


.title-input {
  width: 221px;
}
.label-container{
    border-right: gray;
    border-right-style: solid;
    border-right-width: 1px;
    padding: 10px;
    width: 60%;
}
/* Default styles */

/* Styles for screens with a maximum width of 768px */
@media only screen and (max-width: 768px) {
  /* Styles for this screen size */
  .retro-form-body {
    display: flex;
    flex-direction: column;
    padding: 17px;
    overflow: scroll;
    height: 400px;
  }
  .label-container {
    margin-right: 0px;
    border: 0px;
  }
  .upload-image {
    display: flex;
    gap: 10px;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 252px;
    width: 168%;
    padding: 20px;
    border-radius: 5px;
    border: 2px dashed #555;
    color: #444;
    cursor: pointer;
    transition: 0.2s ease-in-out, border 0.2s ease-in-out;
    margin: 0px;
    margin-top: 10px;
  }
  .upload-images {
    display: flex;
    flex-direction: row;
    text-align: left;
  }
  .label-container label {
    font-size: 20px;
    font-weight: 500;
    color: grey;
    padding-left: 53px;
    margin: 0px;
  }
  .submit-btn {
    margin-top: 10px;
    cursor: pointer;
    font-weight: 500;
    padding: 11px 28px;
    margin-left: 94px;
    border-radius: 5px 5px;
    font-size: 0.8rem;
    border: none;
    color: #fff;
    background: #051c2c;
    margin-right: 10px;
    width: 250px;
  }
  .retro-form-body-right h1 {
    font-size: 22px;
    font-weight: 10px;
    padding-left: 79px;
  }
  .retro-form-body-right button {
    font-weight: 500;
    color: #030c82;
    padding-top: 8px;
    margin-left: 63px;
    padding-left: 40px;
  }
  .action-list {
    list-style: circle;
    list-style-type: disc;
    cursor: pointer;
    font-size: 15px;
    margin-left: 61px;
  }
  .action-list :hover {
    color: #f24727;
  }
  .retro-form-body-right input {
    font-weight: 500;
    color: rgb(43, 43, 111);
    margin-top: 20px;
    margin-left: 80px;
    border-radius: 5px;
  }
  .action-item-input {
    border: 1.5px;
    border-color: blue;
    border-style: solid;
    margin-left: 20px;
    margin-right: 10px;
    width: 35%;
    height: 35px;
    padding-left: 10px;
  }
}

/* Styles for screens with a minimum width of 768px and a maximum width of 992px */
@media only screen and (min-width: 768px) and (max-width: 992px) {
  /* Styles for this screen size */
  .upload-image {
    display: flex;
    gap: 10px;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 327px;
    width: 94%;
    padding: 20px;
    border-radius: 5px;
    border: 2px dashed #555;
    color: #444;
    cursor: pointer;
    transition: 0.2s ease-in-out, border 0.2s ease-in-out;
    margin: 0px;
    margin-top: 10px;
  }
}

/* Styles for screens with a minimum width of 992px and a maximum width of 1200px */
@media only screen and (min-width: 992px) and (max-width: 1200px) {
  /* Styles for this screen size */
}

/* Styles for screens with a minimum width of 1200px */
@media only screen and (min-width: 1200px) {
  /* Styles for this screen size */
  /* .retro-button {
        margin-top: 55px;
        width: 104px;
    }
    .icon {
        font-size: 50px;
        position: relative;
        color: grey;
        bottom: 10px;
        right: 14px;
    } */
}