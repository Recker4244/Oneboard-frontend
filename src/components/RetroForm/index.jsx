/* eslint-disable max-len */
/* eslint-disable jsx-a11y/control-has-associated-label */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, { useState, useEffect } from 'react';
import FileUploadIcon from '@mui/icons-material/FileUpload';
import CloseRoundedIcon from '@mui/icons-material/CloseRounded';
import './RetroForm.css';
import axios from 'axios';
import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';

function RetroForm({ click, setClick }) {
  const [file, setFile] = useState();
  const [submit, setSubmit] = useState(false);
  const [name, setName] = useState('');
  const [actions, setAction] = useState([]);
  const [title, setTitle] = useState('');
  const [buttonclick, setButtonClick] = useState(false);
  const [retroTitle, setRetroTitle] = useState('');
  const { id } = useParams();
  const handleTitleInput = (e) => {
    e.preventDefault();
    setRetroTitle('');
    setTitle(e.target.value);
  };

  const handleFileInput = (e) => {
    e.preventDefault();
    setFile(e.target.files[0]);
  };

  const handleClickButton = () => {
    setClick(!click);
  };
  const handleClick = () => {
    setRetroTitle(title);
    setSubmit(true);
  };
  useEffect(() => {
    const handleSubmit = async () => {
      const data = new FormData();
      data.append('image', file);
      console.log(actions);
      data.append('project_id', `${id}`);
      data.append('title', retroTitle);
      actions.forEach((action) => {
        data.append('action_items', action.name);
      });
      // console.log(data);

      const config = {
        method: 'post',
        url: 'http://localhost:8080/retro/',
        data,
        headers: {
          'Content-Type': 'multipart/form-data',
          'x-auth-token': localStorage.getItem('accessToken')
        }
      };

      try {
        const response = await axios(config);
        console.log(JSON.stringify(response.data));
        window.location.reload();
      } catch (error) {
        console.log(error.message);
      }
    };

    if (submit) {
      handleSubmit();
    }
    // setFile();
  }, [submit, file, retroTitle, actions]);

  const deleteEntry = (actionName) => {
    const newAction = actions.filter((item) => item.name !== actionName);
    setAction(newAction);
  };

  return (
    <div className="retro-form z-50">
      <div className="retro-form-content">
        <div className="retro-form-title">
          <div className="retro-form-title-left">
            <input
              type="text"
              placeholder="Retro Board Title"
              onChange={handleTitleInput}
              className="title-input"
              data-testid="retro-title"
            />
          </div>
          <button type="button" className="cancel-button" onClick={handleClickButton} data-testid="cancel-button"><CloseRoundedIcon sx={{ fontSize: 40 }} /></button>

        </div>
        <div className="retro-form-body">
          <div className="label-container">
            <label htmlFor="input-tag" className="upload-image">
              {!file && (
              <div className="upload-images">
                <FileUploadIcon fontSize="large" />
                <h2 data-testid="retro-form">Upload Retro Board Images</h2>

                <input type="file" id="input-tag" onChange={handleFileInput} />
              </div>
              )}

              {/* create a preview for input image */}
              {file && (
                <div className="re-upload-image">
                  <img src={URL.createObjectURL(file)} alt="preview" />
                  <div className="re-upload">
                    <FileUploadIcon fontSize="large" />
                    <h3>Re-upload Retro Board Image</h3>
                  </div>
                  <input type="file" id="input-tag" onChange={handleFileInput} />
                </div>
              )}
            </label>
          </div>
          {/* <input type="file" className="retro-from-body" onChange={handleFileInput} /> */}
          <div className="retro-form-body-right">
            <h1>ACTION ITEMS</h1>
            {buttonclick && (
            <input
              value={name}
              onChange={(e) => setName(e.target.value)}
              placeholder="Action Items"
              // data-testid="action-item-update"
              onKeyPress={(e) => {
                if (e.key === 'Enter' && name.trim().length > 0) {
                  // setName('');
                  setButtonClick(!buttonclick);
                  setAction((item) => [
                    ...item,
                    {
                      // id: nextId + 1,
                      name
                    }
                  ]);
                }
              }}
              // placeholder="Action Items"
              className="action-item-input"
            />
            )}
            {!buttonclick && (
            <button
              type="submit"
              onClick={() => {
                setName('');
                setButtonClick(!buttonclick);
              }}
            >
              + Add an item
            </button>
            )}
            <ul className="action-lists" data-testId="action-lists">
              {actions.map((action) => (
                // eslint-disable-next-line no-alert
                <li key={action.id} className="action-list" onClick={() => deleteEntry(action.name)} data-testId="delete-action-item">
                  {action.name}
                </li>
              ))}
            </ul>
          </div>
        </div>
        <div className="flex justify-end">
          <button type="submit" className="rounded-3xl border-2 text-sm bg-button-hover text-white hover:font-bold w-36 h-10 mr-10" onClick={handleClick} data-TestId="submit-button">
            Submit
          </button>
        </div>

      </div>
    </div>

  );
}

RetroForm.propTypes = {
  click: PropTypes.bool.isRequired,
  setClick: PropTypes.func.isRequired
};

export default RetroForm;
