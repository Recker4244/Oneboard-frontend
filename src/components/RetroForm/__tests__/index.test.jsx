/* eslint-disable react/react-in-jsx-scope */
import { render, fireEvent } from '@testing-library/react';
import RetroForm from '../index';
import '@testing-library/jest-dom';

describe('RetroForm', () => {
  it('renders RetroForm without crashing', () => {
    render(<RetroForm />);
  });
  it('should display upload image section', () => {
    const { getByTestId } = render(<RetroForm />);
    expect(getByTestId('retro-form')).toHaveTextContent('Upload Retro Board Images');
  });
  it('should display title input', () => {
    const { getByTestId } = render(<RetroForm />);
    const titleInput = getByTestId('retro-title');
    expect(titleInput).toHaveAttribute('placeholder', 'Retro Board Title');
    fireEvent.change(titleInput, { target: { value: 'My Retro Board' } });
    expect(titleInput.value).toBe('My Retro Board');
  });
  //   it('should display action item input', () => {
  //     const { getByTestId } = render(<RetroForm />);
  //     const actionItemInput = getByTestId('action-item-input');
  //     expect(actionItemInput).toHaveAttribute('placeholder', 'Action Item');
  //     fireEvent.change(actionItemInput, { target: { value: 'My Action Item' } });
  //     expect(actionItemInput.value).toBe('My Action Item');
  //   });

  it('should submit the form when clicking the submit button', () => {
    const { getByTestId } = render(<RetroForm />);
    const submitButton = getByTestId('submit-button');
    fireEvent.click(submitButton);
    expect(submitButton).toHaveTextContent('Submit');
  });
  // snapshot test
  it('should match the snapshot', () => {
    const { asFragment } = render(<RetroForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  //   it('should cancel the form when clicking the cancel button', () => {
  //     const { getByTestId } = render(<RetroForm />);
  //     const cancelButton = getByTestId('cancel-button');
  //     fireEvent.click(cancelButton);
  //     expect(cancelButton).toBeTruthy();
  //   });
  // it('should delete the action item when clicking on it', () => {
  //   const deleteEntry = jest.fn();

  //   // Render the component with a test action item
  //   const { getAllByTestId, getByTestId } = render(
  //     <RetroForm actions={[{ id: 1, name: 'Test action item' }]} deleteEntry={deleteEntry} />
  //   );

  //   // Find and click the action item
  //   const actionItem = getAllByTestId('action-list')[0]);
  //   fireEvent.click(actionItem);
  //   // console.log(actionItem);
  //   // Verify that deleteEntry has been called with the correct argument
  //   expect(deleteEntry).toHaveBeenCalledWith('Test action item');

  //   // Verify that the action item has been removed from the state
  //   const updatedActions = getByTestId('action-lists').children;
  //   expect(updatedActions).toHaveLength(0);
  // });

  it('should reset name and toggle button click state when the add item button is clicked', () => {
    // Mock setName and setButtonClick
    const setName = jest.fn();
    const setButtonClick = jest.fn();

    // Render the component with the mock functions
    const { getByText } = render(<RetroForm setName={setName} setButtonClick={setButtonClick} />);

    // Find the add item button and click it
    const addItemButton = getByText('+ Add an item');
    fireEvent.click(addItemButton);

    // Check that setName and setButtonClick have been called with the correct arguments
    expect(setName).toHaveBeenCalledWith('');
    expect(setButtonClick).toHaveBeenCalledWith(true);
  });
});
