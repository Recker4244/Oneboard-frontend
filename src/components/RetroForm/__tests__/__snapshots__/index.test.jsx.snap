// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RetroForm should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="retro-form z-50"
  >
    <div
      class="retro-form-content"
    >
      <div
        class="retro-form-title"
      >
        <div
          class="retro-form-title-left"
        >
          <input
            class="title-input"
            data-testid="retro-title"
            placeholder="Retro Board Title"
            type="text"
          />
        </div>
        <button
          class="cancel-button"
          data-testid="cancel-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ha5wk2-MuiSvgIcon-root"
            data-testid="CloseRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18.3 5.71a.9959.9959 0 0 0-1.41 0L12 10.59 7.11 5.7a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
            />
          </svg>
        </button>
      </div>
      <div
        class="retro-form-body"
      >
        <div
          class="label-container"
        >
          <label
            class="upload-image"
            for="input-tag"
          >
            <div
              class="upload-images"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                data-testid="FileUploadIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
                />
              </svg>
              <h2
                data-testid="retro-form"
              >
                Upload Retro Board Images
              </h2>
              <input
                id="input-tag"
                type="file"
              />
            </div>
          </label>
        </div>
        <div
          class="retro-form-body-right"
        >
          <h1>
            ACTION ITEMS
          </h1>
          <button
            type="submit"
          >
            + Add an item
          </button>
          <ul
            class="action-lists"
            data-testid="action-lists"
          />
        </div>
      </div>
      <div
        class="flex justify-end"
      >
        <button
          class="rounded-3xl border-2 text-sm bg-button-hover text-white hover:font-bold w-36 h-10 mr-10"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
